File: mondrian-charts/src/generate_shape_list_from_events.rs
use super::utils::*;
use crate::types::{
    Area, AreaPoint, Axis, PointSource, ProviderEvent, SeriesSource, Shape, ShapeList,
};

pub(crate) fn generate_shape_list_from_events<'source>(
    axis: &Axis,
    events: &[&'source ProviderEvent],
) -> ShapeList<SeriesSource<'source>, PointSource<'source>> {
    ShapeList {
        shapes: events
            .iter()
            .map(|event| {
                Shape::Area(Area {
                    points: vec![AreaPoint {
                        x: normalize_along_linear_axis(get_time_from_timestamp(event.time), axis),
                        y_min: 0.,
                        y_max: 1.,
                        source: PointSource::Event(event),
                    }],
                })
            })
            .collect(),
        source: SeriesSource::Events,
    }
}

File: fiberplane-charts/src/Mondrian/generator/generateShapeListFromEvents.ts
import type { ProviderEvent } from "../../providerTypes";
import type { Axis, SeriesSource, ShapeList } from "../types";
import { getTimeFromTimestamp } from "./utils";

export function generateShapeListFromEvents(
  { minValue, maxValue }: Axis,
  events: Array<ProviderEvent>,
): ShapeList<SeriesSource, ProviderEvent> {
  return {
    shapes: events.map((event) => {
      const x =
        (getTimeFromTimestamp(event.time) - minValue) / (maxValue - minValue);
      return {
        type: "area",
        points: [{ x, yMin: 0, yMax: 1, source: event }],
        strokeDasharray: [2],
      };
    }),
    source: { type: "events" },
  };
}
